lib/container/HoistContainerReferencesPlugin.js

```diff
@@ -0,0 +1,205 @@
+			}
+
+			const containerRuntimes =
+				chunkGraph.getModuleRuntimes(containerEntryModule);
+			const runtimes = new Set();
+
+			for (const runtimeSpec of containerRuntimes) {
+				compilation.compiler.webpack.util.runtime.forEachRuntime(
+					runtimeSpec,
+					runtimeKey => {
+						if (runtimeKey) {
+							runtimes.add(runtimeKey);
+						}
+					}
+				);
+			}
+
+			for (const runtime of runtimes) {
+				const runtimeChunk = compilation.namedChunks.get(runtime);
+				if (!runtimeChunk) continue;
+
+				for (const module of allReferencedModules) {
+					if (!chunkGraph.isModuleInChunk(module, runtimeChunk)) {
+						chunkGraph.connectChunkAndModule(runtimeChunk, module);
+					}
+				}
+			}
+			this.cleanUpChunks(compilation, allReferencedModules);
+		}
+	}
+
+	/**
+	 * Clean up chunks by disconnecting unused modules.
+	 * @param {import("webpack").Compilation} compilation The webpack compilation instance.
+	 * @param {Set<import("webpack").Module>} modules Set of modules to clean up.
+	 */
+	cleanUpChunks(compilation, modules) {
+		const { chunkGraph } = compilation;
+		for (const module of modules) {
+			for (const chunk of chunkGraph.getModuleChunks(module)) {
+				if (!chunk.hasRuntime()) {
+					chunkGraph.disconnectChunkAndModule(chunk, module);
+					if (
+						chunkGraph.getNumberOfChunkModules(chunk) === 0 &&
+						chunkGraph.getNumberOfEntryModules(chunk) === 0
+					) {
+						chunkGraph.disconnectChunk(chunk);
+						compilation.chunks.delete(chunk);
+						if (chunk.name) {
+							compilation.namedChunks.delete(chunk.name);
+						}
+					}
+				}
+			}
+		}
+		modules.clear();
+	}
+
+	/**
+	 * Helper method to get runtime chunks from the compilation.
+	 * @param {import("webpack").Compilation} compilation The webpack compilation instance.
+	 * @returns {Set<import("webpack").Chunk>} Set of runtime chunks.
+	 */
+	getRuntimeChunks(compilation) {
+		const runtimeChunks = new Set();
+		const entries = compilation.entrypoints;
+
+		for (const entrypoint of entries.values()) {
+			const runtimeChunk = entrypoint.getRuntimeChunk();
+			if (runtimeChunk) {
+				runtimeChunks.add(runtimeChunk);
+			}
+		}
+		return runtimeChunks;
+	}
+}
+
+/**
+ * Helper method to collect all referenced modules recursively.
+ * @param {import("webpack").Compilation} compilation The webpack compilation instance.
+ * @param {import("webpack").Module} module The module to start collecting from.
+ * @param {string} type The type of modules to collect ("initial", "external", or "all").
+ * @returns {Set<import("webpack").Module>} Set of collected modules.
+ */
+function getAllReferencedModules(compilation, module, type) {
+	const collectedModules = new Set([module]);
+	const visitedModules = new WeakSet([module]);
+	const stack = [module];
+
+	while (stack.length > 0) {
+		const currentModule = stack.pop();
+		if (!currentModule) continue;
+
+		const mgm = compilation.moduleGraph._getModuleGraphModule(currentModule);
```

Comment:
gives lint / type error but not sure how else to get the modulegraph connections in this way.

Reply:
done